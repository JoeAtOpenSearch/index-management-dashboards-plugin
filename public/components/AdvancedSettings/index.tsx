import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef } from "react";
import { EuiAccordion, EuiAccordionProps, EuiFormRow, EuiSpacer, EuiFormRowProps } from "@elastic/eui";
import JSONEditor, { IJSONEditorRef, JSONEditorProps } from "../JSONEditor";
import CustomFormRow from "../CustomFormRow";
import "./index.scss";
import CustomFormRow from "../CustomFormRow";

export interface IAdvancedSettingsProps {
  rowProps?: Omit<EuiFormRowProps, "children">;
  accordionProps?: EuiAccordionProps;
  value?: Record<string, any>;
  onChange?: (totalValue: IAdvancedSettingsProps["value"]) => void;
  renderProps?: (
    options: Pick<Required<IAdvancedSettingsProps>, "value" | "onChange"> & { ref: React.Ref<IJSONEditorRef> }
  ) => React.ReactChild;
  editorProps?: Partial<JSONEditorProps>;
}

export interface IAdvancedSettingsRef {
  validate: () => Promise<string>;
}

export default forwardRef(function AdvancedSettings(props: IAdvancedSettingsProps, ref: React.Ref<IAdvancedSettingsRef>) {
  const { value, renderProps, editorProps } = props;
  const propsRef = useRef<IAdvancedSettingsProps>(props);
  const editorRef = useRef<IJSONEditorRef>(null);
  propsRef.current = props;

  const onChangeInRenderProps = useCallback(
    (val: string) => {
      const parsedValue = JSON.parse(val);
      propsRef.current.onChange && propsRef.current.onChange(parsedValue);
    },
    [props.onChange]
  );

  const accordionId = useMemo(() => {
    return props.accordionProps?.id || `${Date.now()}-${Math.floor(Math.random() * 100)}`;
  }, [props.accordionProps?.id]);

  useImperativeHandle(ref, () => ({
    validate: () => {
      if (editorRef.current) {
        return editorRef.current.validate();
      }

      return Promise.resolve("");
    },
  }));

  return (
    <>
      <EuiSpacer size="m" />
      <EuiAccordion {...props.accordionProps} className="accordion-in-advanced-settings" id={accordionId}>
        <CustomFormRow {...(props.rowProps as EuiFormRowProps)}>
          {renderProps ? (
            (renderProps({
              value: propsRef.current.value || {},
              onChange: (val) => {
                if (propsRef.current?.onChange) {
                  propsRef.current?.onChange(val);
                }
              },
              ref: editorRef,
            }) as any)
          ) : (
            <JSONEditor {...editorProps} ref={editorRef} value={JSON.stringify(value, null, 2)} onChange={onChangeInRenderProps} />
          )}
        </CustomFormRow>
      </EuiAccordion>
    </>
  );
});
