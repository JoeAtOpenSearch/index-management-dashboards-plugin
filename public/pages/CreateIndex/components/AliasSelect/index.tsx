/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { forwardRef } from "react";
import { EuiComboBoxProps } from "@elastic/eui";
import RemoteSelect, { RemoteSelectProps } from "../../../../components/RemoteSelect";
import { ServerResponse } from "../../../../../server/models/types";

export interface AliasSelectProps extends Omit<EuiComboBoxProps<{ label: string; value: string }>, "value" | "onChange"> {
  value?: Record<string, {}>;
  onChange?: (value: AliasSelectProps["value"]) => void;
  refreshOptions: (aliasName: string) => Promise<ServerResponse<{ alias: string }[]>>;
}

const transformObjToArray = (obj: AliasSelectProps["value"]): { label: string }[] => {
  return Object.keys(obj || {}).map((label) => ({ label }));
};
const transformArrayToObj = (array: { label: string }[]): AliasSelectProps["value"] => {
  return array.reduce((total, current) => ({ ...total, [current.label]: {} }), {});
};

const AliasSelect = forwardRef((props: AliasSelectProps, ref: React.Ref<HTMLInputElement>) => {
  const { value, onChange, refreshOptions: refreshOptionsFromProps } = props;
  const refreshOptions: RemoteSelectProps["refreshOptions"] = ({ searchValue }) => {
    return refreshOptionsFromProps(searchValue || "").then((res) => {
      if (res.ok) {
        return {
          ...res,
          response: [...new Set(res.response.map((item) => item.alias).filter((item) => !item.startsWith(".")))].map((item) => ({
            label: item,
          })),
        };
      } else {
        return res;
      }
    });
  };
  return (
    <RemoteSelect
      isDisabled={props.isDisabled}
      placeholder="Select or create aliases"
      customOptionText="Add {searchValue} as a new alias"
      refreshOptions={refreshOptions}
      value={transformObjToArray(value).map((item) => item.label)}
      onChange={(val) => {
        onChange && onChange(transformArrayToObj(val.map((label) => ({ label }))));
      }}
    />
  );
});

AliasSelect.displayName = "AliasSelect";

// @ts-ignore
export default AliasSelect;
