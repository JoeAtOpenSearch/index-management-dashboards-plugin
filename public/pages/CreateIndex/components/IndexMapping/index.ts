/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexMapping from "./IndexMapping";

export default IndexMapping;
