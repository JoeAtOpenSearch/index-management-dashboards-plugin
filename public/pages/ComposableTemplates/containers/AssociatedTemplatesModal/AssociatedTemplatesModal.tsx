/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Dispatch, SetStateAction, useContext, useState } from "react";
import { CoreStart } from "opensearch-dashboards/public";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";
import { EuiButtonIcon, EuiFlyout, EuiFlyoutBody, EuiFlyoutHeader, EuiInMemoryTable, EuiLink, EuiTitle } from "@elastic/eui";
import { ROUTES } from "../../../../utils/constants";
import { ReactChild } from "react";
import { Modal } from "../../../../components/Modal";
import { getTemplate } from "../../utils/hooks";
import { BrowserServices } from "../../../../models/interfaces";
import { TemplateItemRemote } from "../../../../../models/interfaces";

interface AssociatedTemplatesModalProps {
  componentTemplate: string;
  associatedTemplates: string[];
  onUnlink: (unlinkTemplate: string) => void;
  renderProps: (params: { setVisible: Dispatch<SetStateAction<boolean>> }) => ReactChild;
}

export default function AssociatedTemplatesModalProps(props: AssociatedTemplatesModalProps) {
  const { onUnlink, associatedTemplates, renderProps, componentTemplate } = props;
  const [visible, setVisible] = useState(false);
  const services = useContext(ServicesContext) as BrowserServices;
  const coreServices = useContext(CoreServicesContext) as CoreStart;

  return (
    <>
      {renderProps ? renderProps({ setVisible }) : null}
      {visible ? (
        <EuiFlyout onClose={() => setVisible(false)}>
          <EuiFlyoutHeader>
            <EuiTitle>
              <h2>Associated templates</h2>
            </EuiTitle>
          </EuiFlyoutHeader>
          <EuiFlyoutBody>
            <EuiInMemoryTable
              items={associatedTemplates.map((item) => ({
                name: item,
              }))}
              columns={[
                {
                  name: "Templates",
                  field: "name",
                  sortable: true,
                  render: (value: string, record) => (
                    <EuiLink external={false} target="_blank" href={`#${ROUTES.CREATE_TEMPLATE}/${value}/readonly`}>
                      {value}
                    </EuiLink>
                  ),
                },
                {
                  name: "Actions",
                  field: "actions",
                  render: (value: string, record) => {
                    return (
                      <EuiButtonIcon
                        aria-label={`Unlink from ${record.name}?`}
                        iconType="unlink"
                        onClick={() => {
                          Modal.show({
                            type: "confirm",
                            title: `Unlink from ${record.name}?`,
                            content: `The component ${componentTemplate} will be removed from the template ${record.name}. This will affect any new indexes that will be created with the template.`,
                            footer: ["cancel", "confirm"],
                            locale: {
                              ok: "Unlink",
                            },
                            async onOk() {
                              const currentTemplate = await getTemplate({
                                templateName: record.name,
                                commonService: services.commonService,
                                coreService: coreServices,
                              });
                              const updateResult = await services.commonService.apiCaller({
                                endpoint: "transport.request",
                                data: {
                                  method: "POST",
                                  path: `_index_template/${record.name}`,
                                  body: {
                                    ...currentTemplate,
                                    composed_of: currentTemplate.composed_of?.filter((item) => item !== componentTemplate) || [],
                                  } as TemplateItemRemote,
                                },
                              });
                              if (updateResult.ok) {
                                onUnlink(record.name);
                                coreServices.notifications.toasts.addSuccess(
                                  `${componentTemplate} has been successfully unlinked from ${record.name}.`
                                );
                              } else {
                                coreServices.notifications.toasts.addDanger(updateResult.error);
                              }
                            },
                          });
                        }}
                      />
                    );
                  },
                },
              ]}
            />
          </EuiFlyoutBody>
        </EuiFlyout>
      ) : null}
    </>
  );
}
