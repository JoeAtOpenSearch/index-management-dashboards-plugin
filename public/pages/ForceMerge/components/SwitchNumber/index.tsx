import React, { useState } from "react";
import { EuiRadioGroup, EuiSpacer } from "@elastic/eui";
import CustomFormRow from "../../../../components/CustomFormRow";
import { AllBuiltInComponents } from "../../../../components/FormGenerator";

interface SwitchNumberProps {
  value?: number;
  onChange: (val: SwitchNumberProps["value"]) => void;
}

export default function SwitchNumber(props: SwitchNumberProps) {
  const [id, setId] = useState(props.value && props.value > 0 ? "1" : "0");
  return (
    <>
      <EuiRadioGroup
        options={[
          {
            id: "0",
            label: "Automatically",
          },
          {
            id: "1",
            label: "Manually set",
          },
        ]}
        idSelected={id}
        onChange={(id) => {
          setId(id);
          if (id === "0") {
            props.onChange(undefined);
          } else {
            props.onChange(1);
          }
        }}
      />
      {id === "1" ? (
        <>
          <EuiSpacer size="s" />
          <CustomFormRow helpText="Specify the number of segments to merge to">
            <AllBuiltInComponents.Number {...props} />
          </CustomFormRow>
        </>
      ) : null}
    </>
  );
}
