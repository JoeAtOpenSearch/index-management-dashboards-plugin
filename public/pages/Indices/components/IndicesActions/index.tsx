/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useCallback, useContext, useMemo, useState } from "react";
import {
  // @ts-ignore
  Criteria,
  // @ts-ignore
  Pagination,
  EuiButton,
  EuiContextMenu,
} from "@elastic/eui";

import { ManagedCatIndex } from "../../../../../server/models/interfaces";
import { ServerResponse } from "../../../../../server/models/types";
import ApplyPolicyModal from "../../components/ApplyPolicyModal";
import SimplePopover from "../../../../components/SimplePopover";
import { ModalConsumer } from "../../../../components/Modal";
import { CoreServicesContext } from "../../../../components/core_services";
import DeleteIndexModal from "../../components/DeleteIndexModal";
import { ServicesContext } from "../../../../services";

export interface IndicesActionsProps {
  selectedItems: ManagedCatIndex[];
  onDelete: (indices: string) => Promise<ServerResponse<any>>;
}

export default function IndicesActions(props: IndicesActionsProps) {
  const { selectedItems, onDelete } = props;
  const [deleteIndexModalVisible, setDeleteIndexModalVisible] = useState(false);
  const context = useContext(CoreServicesContext);
  const services = useContext(ServicesContext);

  const onDeleteIndexModalClose = () => {
    setDeleteIndexModalVisible(false);
  };

  const onDeleteIndexModalConfirm = useCallback(async () => {
    const result = await onDelete(selectedItems.map((item) => item.index).join(","));
    if (result && result.ok) {
      onDeleteIndexModalClose();
    }
  }, [selectedItems, onDelete, services, context, onDeleteIndexModalClose]);

  const renderKey = useMemo(() => Date.now(), [selectedItems]);

  return (
    <>
      <ModalConsumer>
        {({ onShow }) => (
          <SimplePopover
            data-test-subj="More Action"
            panelPaddingSize="none"
            button={
              <EuiButton iconType="arrowDown" iconSide="right">
                Actions
              </EuiButton>
            }
          >
            <EuiContextMenu
              initialPanelId={0}
              // The EuiContextMenu has bug when testing in jest
              // the props change won't make it rerender
              key={renderKey}
              panels={[
                {
                  id: 0,
                  items: [
                    {
                      name: "Delete",
                      disabled: !selectedItems.length,
                      "data-test-subj": "Delete Action",
                      onClick: () => setDeleteIndexModalVisible(true),
                    },
                    {
                      name: "Apply policy",
                      disabled: !selectedItems.length,
                      "data-test-subj": "Apply policyButton",
                      onClick: () =>
                        onShow(ApplyPolicyModal, {
                          indices: selectedItems.map((item: ManagedCatIndex) => item.index),
                          core: context,
                        }),
                    },
                  ],
                },
              ]}
            />
          </SimplePopover>
        )}
      </ModalConsumer>
      <DeleteIndexModal
        selectedItems={selectedItems.map((item) => item.index)}
        visible={deleteIndexModalVisible}
        onClose={onDeleteIndexModalClose}
        onConfirm={onDeleteIndexModalConfirm}
      />
    </>
  );
}
