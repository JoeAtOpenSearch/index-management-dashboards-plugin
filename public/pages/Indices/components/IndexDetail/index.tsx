/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from "react";
import { EuiButtonEmpty, EuiCopy } from "@elastic/eui";
import { IndexItem } from "../../../../../models/interfaces";
import { Modal } from "../../../../components/Modal";
import IndicesActions from "../IndicesActions";
import { ManagedCatIndex } from "../../../../../server/models/interfaces";

export interface IndexDetailModalProps {
  index: string;
  onDelete?: () => void;
  record: ManagedCatIndex;
}

export default function IndexDetail(props: IndexDetailModalProps) {
  // const { detail, visible } = props;
  const { index, record, onDelete } = props;
  const [visible, setVisible] = useState(false);
  const [detail, setDetail] = useState({} as IndexItem);
  return (
    <>
      <EuiCopy textToCopy={index}>
        {(copy) => <EuiButtonEmpty size="xs" flush="right" iconType="copyClipboard" onClick={copy} color="text"></EuiButtonEmpty>}
      </EuiCopy>
      <EuiButtonEmpty onClick={() => setVisible(true)} data-test-subj={`view-index-detail-button-${index}`}>
        {index}
      </EuiButtonEmpty>
      <Modal.SimpleModal
        visible={visible}
        maxWidth={false}
        style={{
          width: "60vw",
        }}
        onClose={() => setVisible(false)}
        title={
          <div style={{ display: "flex", justifyContent: "space-between" }}>
            <span>{index}</span>
            <IndicesActions selectedItems={[record]} onDelete={onDelete} />
          </div>
        }
      />
    </>
  );
}
