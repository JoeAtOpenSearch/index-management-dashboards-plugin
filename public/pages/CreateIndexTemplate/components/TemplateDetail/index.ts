/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexDetail from "./TemplateDetail";

export default IndexDetail;
export * from "./TemplateDetail";
