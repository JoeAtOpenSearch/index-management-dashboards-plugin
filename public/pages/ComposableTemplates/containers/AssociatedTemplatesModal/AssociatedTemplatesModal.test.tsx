/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import AssociatedTemplatesModal from "./AssociatedTemplatesModal";
import userEvent from "@testing-library/user-event";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";

describe("<AssociatedTemplatesModal /> spec", () => {
  it("renders the component", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(
      async (payload): Promise<any> => {
        if (payload.endpoint === "transport.request" && payload.data.path === "/_index_template/*") {
          return {
            ok: true,
            response: {
              index_templates: [],
            },
          };
        }

        return {
          ok: true,
        };
      }
    );
    const { findByText, getByTestId } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ServicesContext.Provider value={browserServicesMock}>
          <AssociatedTemplatesModal
            componentTemplate="test"
            renderProps={({ setVisible }) => (
              <div data-test-subj="test" onClick={() => setVisible(true)}>
                123
              </div>
            )}
          />
        </ServicesContext.Provider>
      </CoreServicesContext.Provider>
    );
    await findByText("123");
    await userEvent.click(getByTestId("test"));
    await findByText("Associated index templates");
    await findByText(/This table contains 0 rows;/);
    expect(document.body.children).toMatchSnapshot();
  });
});
