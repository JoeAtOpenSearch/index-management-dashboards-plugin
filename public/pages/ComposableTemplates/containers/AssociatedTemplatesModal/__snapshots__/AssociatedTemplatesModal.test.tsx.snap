// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssociatedTemplatesModal /> spec renders the component 1`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <div
      data-test-subj="test"
    >
      123
    </div>
  </div>,
  <div
    class="euiOverlayMask euiOverlayMask--belowHeader"
  >
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="euiFlyout euiFlyout--medium euiFlyout--paddingLarge"
        role="dialog"
        tabindex="-1"
      >
        <button
          aria-label="Close this dialog"
          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
          data-test-subj="euiFlyoutCloseButton"
          type="button"
        >
          EuiIconMock
        </button>
        <div
          class="euiFlyoutHeader"
        >
          <h2
            class="euiTitle euiTitle--medium"
          >
            Associated index templates
          </h2>
        </div>
        <div
          class="euiFlyoutBody"
        >
          <div
            class="euiFlyoutBody__overflow"
            tabindex="0"
          >
            <div
              class="euiFlyoutBody__overflowContent"
            >
              <div
                class="euiBasicTable"
              >
                <div>
                  <div
                    class="euiTableHeaderMobile"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                    </div>
                  </div>
                  <table
                    class="euiTable euiTable--responsive"
                    id="some_html_id"
                    tabindex="-1"
                  >
                    <caption
                      class="euiScreenReaderOnly euiTableCaption"
                    />
                    <thead>
                      <tr>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_name_0"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Templates"
                            >
                              Templates
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_actions_1"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent euiTableCellContent--alignRight"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Actions"
                            >
                              Actions
                            </span>
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        class="euiTableRow"
                      >
                        <td
                          class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                          colspan="2"
                        >
                          <div
                            class="euiTableCellContent euiTableCellContent--alignCenter"
                          >
                            <span
                              class="euiTableCellContent__text"
                            >
                              No items found
                            </span>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;
